packages/arb-validator-core/gobridge/goEth.go

```diff
@@ -0,0 +1,370 @@
+	gracePeriod            uint64
+	maxSteps               uint32
+	escrowRequired         *big.Int
+	owner                  common.Address
+	ValidatorCount         int64
+	ActiveValidators       int64
+	validatorKeys          []common.Address
+	MachineHash            [32]byte
+	PendingHash            [32]byte
+	Inbox                  [32]byte
+	Asserter               common.Address
+	Deadline               uint64
+	SequenceNum            uint64
+	ActiveChallengeManager common.Address
+	ValidatorBalances      map[common.Address]*big.Int
+}
+
+type rollupData struct {
+	initVMHash              common.Hash
+	VMstate                 machine.Status
+	state                   EthState
+	gracePeriod             common.TimeTicks
+	maxSteps                uint64
+	maxTimeBoundsWidth      uint64
+	arbGasSpeedLimitPerTick uint64
+	escrowRequired          *big.Int
+	owner                   common.Address
+	events                  map[*common.BlockId][]arbbridge.Event
+	creation                *common.BlockId
+	stakers                 map[common.Address]*staker
+	leaves                  map[common.Hash]bool
+	lastConfirmed           common.Hash
+	contractAddress         common.Address
+	nextConfirmed           common.Hash
+}
+
+type challengeData struct {
+	sync.Mutex
+	deadline             common.TimeTicks
+	challengerDataHash   common.Hash
+	state                int
+	challengePeriodTicks common.TimeTicks
+	asserter             common.Address
+	challenger           common.Address
+	challengeType        *big.Int
+}
+
+type pendingTopChallengeData struct {
+	challengeData
+}
+
+type messagesChallengeData struct {
+	challengeData
+}
+
+type execChallengeData struct {
+	challengeData
+}
+
+type staker struct {
+	location           common.Hash
+	creationTimeBlocks *common.TimeBlocks
+	inChallenge        bool
+	balance            *big.Int
+}
+
+type nodeGraph struct {
+	stakeRequirement *big.Int
+	stakers          map[common.Address]*staker
+	leaves           map[common.Hash]bool
+	lastConfirmed    common.Hash
+}
+
+type goMaybeEvent struct {
+	challenge *challengeData
+	event     arbbridge.MaybeEvent
+}
+
+type inbox struct {
+	value common.Hash
+	count *big.Int
+}
+
+func (ib *inbox) addMessageToInbox(msg common.Hash) {
+	ib.value = hashing.SoliditySHA3(
+		hashing.Bytes32(ib.value),
+		hashing.Bytes32(msg),
+	)
+	ib.count = new(big.Int).Add(ib.count, big.NewInt(1))
+}
+
+type void struct{}
+
+var Void void
+
+// goEthData one per 'URL'
+type goEthdata struct {
+	blockMutex             sync.Mutex
+	msgMutex               sync.Mutex
+	Vm                     map[common.Address]*VmData
+	channels               map[common.Address]*channelData
+	rollups                map[common.Address]*rollupData // contract address to rollup
+	challenges             map[common.Address]*challengeData
+	balances               map[common.Address]*big.Int
+	challengeWatchersMutex sync.Mutex
+	challengeWatcherEvents map[*challengeData]map[*common.BlockId][]arbbridge.Event
```

Comment:
Why is there a separate store for challenge events compared to other events?

Reply:
cleaned up event handling